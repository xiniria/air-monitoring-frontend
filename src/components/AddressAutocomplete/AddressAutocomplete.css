.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 3.5rem 5%;
  text-align: center;
}

input {
  padding: 1rem 1.2rem;
  width: 15rem;
  border: 3px solid mediumseagreen;
  outline: none;
  font-weight: bold;
  line-height: 1.5;
}

ul {
  position: absolute;
  margin: 0;
  padding: 0;
  max-height: 60%;
  border: 3px solid mediumseagreen;
  border-top: none;
  list-style-type: none;
  text-align: left;
  overflow-y: auto;
}

li {
  padding: 1rem 1.2rem;
  width: 15rem;
  cursor: pointer;
}
