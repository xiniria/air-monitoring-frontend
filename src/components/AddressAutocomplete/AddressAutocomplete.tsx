import React from 'react';
import usePlacesAutocomplete, {
  getGeocode,
  getLatLng,
  Suggestion,
} from 'use-places-autocomplete';
import useOnclickOutside from 'react-cool-onclickoutside';
import './AddressAutocomplete.css';
import useGeolocation from 'hooks/useGeolocation/useGeolocation';
import { useLocation } from 'LocationContext';
import { LocationTypes } from 'interfaces/location';

function AddressAutocomplete(): JSX.Element {
  const currentPos = useGeolocation();
  const { setLocation } = useLocation();

  const {
    ready,
    value,
    suggestions: { status, data },
    setValue,
    clearSuggestions,
  } = usePlacesAutocomplete({
    requestOptions: {
      componentRestrictions: { country: 'fr' },
    },
    debounce: 300,
  });

  const ref = useOnclickOutside(() => {
    // When user clicks outside of the component, we can dismiss
    // the searched suggestions by calling this method
    clearSuggestions();
  });

  const handleInput = (e: React.FormEvent<HTMLInputElement>) => {
    // Update the keyword of the input element
    setValue(e.currentTarget.value);
  };

  const getCurrentPosition = () => () => {
    // When user selects a place, we can replace the keyword without request data from API
    // by setting the second parameter as "false"
    clearSuggestions();
    // Get latitude and longitude via utility functions
    setLocation({
      latitude: currentPos.latitude,
      longitude: currentPos.longitude,
      error: currentPos.error,
      type: LocationTypes.Geolocation,
    });
  };

  const handleSelect = ({ description }: Suggestion) => async () => {
    // When user selects a place, we can replace the keyword without request data from API
    // by setting the second parameter as "false"
    setValue(description, false);
    clearSuggestions();

    // Get latitude and longitude via utility functions
    try {
      const results = await getGeocode({ address: description });
      const coords = await getLatLng(results[0]);
      console.log('üìç Coordinates: ', coords);
      setLocation({
        latitude: coords.lat,
        longitude: coords.lng,
        error: '',
        type: LocationTypes.Address,
      });
    } catch (error) {
      setLocation({
        latitude: NaN,
        longitude: NaN,
        error: error,
        type: LocationTypes.Address,
      });
      console.log('üò± Error: ', error);
    }
  };

  const renderSuggestions = () =>
    data.map((suggestion, index) => {
      const {
        id,
        structured_formatting: { main_text, secondary_text },
      } = suggestion;

      return (
        <li
          className="address"
          key={id ? +index : index}
          onClick={handleSelect(suggestion)}
        >
          <strong>{main_text}</strong> <small>{secondary_text}</small>
        </li>
      );
    });

  return (
    <div ref={ref}>
      <input
        className="address-input"
        value={value}
        onChange={handleInput}
        disabled={!ready}
        placeholder="Choisissez un endroit..."
      />
      {/* We can use the "status" to decide whether we should display the dropdown or not */}
      {status === 'OK' && (
        <ul className="address-suggestions">
          <li
            className="address"
            key="currentPosition"
            onClick={getCurrentPosition()}
          >
            <strong>üìç Ma position</strong>
          </li>
          {renderSuggestions()}
        </ul>
      )}
    </div>
  );
}

export default AddressAutocomplete;
