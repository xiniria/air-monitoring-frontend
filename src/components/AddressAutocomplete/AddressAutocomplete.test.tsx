import React from 'react';
import Adapter from  '@wojtekmaj/enzyme-adapter-react-17';
import { configure, shallow } from 'enzyme';
import AddressAutocomplete from 'components/AddressAutocomplete/AddressAutocomplete';

configure({adapter: new Adapter()});

test('renders correctly address input', () => {
  const addressInput = shallow(<AddressAutocomplete />).find(AddressAutocomplete);
  expect(addressInput).toMatchSnapshot();
});

test('displays input', () => {
  const wrapper = shallow(<AddressAutocomplete />);
  expect(wrapper.find('input')).toHaveLength(1);
});

test('calls handleInput on change to set to the new value', () => {
  const wrapper = shallow(<AddressAutocomplete />);
  wrapper
    .find('input')
    .simulate('change', { currentTarget: { value: '1 rue Joliot Curie' } });
  expect(wrapper.find('input').props().value).toContain('1 rue Joliot Curie');
});
