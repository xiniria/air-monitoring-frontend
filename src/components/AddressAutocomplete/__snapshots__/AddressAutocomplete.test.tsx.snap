// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly address input 1`] = `ShallowWrapper {}`;
